libpam/pam_google_authenticator.c

```diff
@@ -350,10 +352,22 @@ static int open_secret_file(pam_handle_t *pamh, const char *secret_filename,
     return -1;
   }
 
+  // print comparisson
+  if (params->debug) {
+    log_message(LOG_INFO, pamh, "Secret file permissions are %4o", orig_stat->st_mode & 03577);
+    if (params->allow_perm) {
+      log_message(LOG_INFO, pamh, "Extra file permissions allowed are %4o", params->allow_perm);
+    }
+  }
+
   // Check permissions on "~/.google_authenticator"
-  if ((orig_stat->st_mode & 03577) != 0400 ||
+  if (((orig_stat->st_mode & 03577) != 0400 &&
```

Comment:
I agree that is more readable, but that has sightly different behaviour than the current change.

Currently, with a setting of `allow_perm=777` then both permissions of exactly 400 and exactly 777 are allowed. With the code above, the permission of 777 would override 400 as the only allowed permission.

A possible middle ground could be something like:
```
int allow_perm = params->allow_perm | 0400;
if ((orig_stat->st_mode & 03577) & allow_perm != 0 ...
```

This would allow file permissions where `allow_perm` is applied as a permission mask. Only permission bits set in the mask (plus the owner-read bit) would be allowed to be set on the file.